src/server/lock_manager/mod.rs

```diff
@@ -564,6 +565,155 @@ mod tests {
+                24,
+                14,
+                b"k8",
+                b"tag9",
+                Some(13),
+                Some(DeadlockInfo {
+                    waiter_ts: 24,
+                    key: b"k7",
+                    wait_chain: &[(13, 24, b"k7", b"tag8"), (24, 13, b"k8", b"tag9")],
+                }),
+            ),
+            // 10->25, change 25 to 26 on k9, updates normally without forming any deadlock.
+            (10, 25, b"k9", b"tag9", Some(26), None),
+            // 27->11, change 11 to 10 on k2, updates normally without forming any deadlock.
+            (27, 11, b"k2", b"taga", Some(10), None),
+        ] {
+            let (waiter, mut lock_info, f) =
+                new_test_waiter_with_key(waiter_ts.into(), lock_ts.into(), key);
+            let token = lock_mgr.allocate_token();
+            lock_mgr.wait_for(
+                token,
+                1,
+                RegionEpoch::default(),
+                1,
+                waiter.start_ts,
+                waiter.wait_info.clone(),
+                false,
+                Some(WaitTimeout::Millis(500)),
+                waiter.cancel_callback,
+                diag_ctx(resource_group_tag, INVALID_TRACKER_TOKEN),
+            );
+
+            if let Some(new_lock_ts) = new_lock_ts {
+                lock_info.lock_version = new_lock_ts;
+                let mut new_wait_info = waiter.wait_info;
+                new_wait_info.lock_digest.ts = new_lock_ts.into();
+                new_wait_info.lock_info = lock_info.clone();
+
+                update_events.push(UpdateWaitForEvent {
+                    token,
+                    start_ts: waiter_ts.into(),
+                    is_first_lock: false,
+                    wait_info: new_wait_info,
+                })
+            }
+            if let Some(deadlock_info) = deadlock_info {
+                deadlock_handles.push((deadlock_info, lock_info, f));
+            } else {
+                handles.push((lock_info, f));
+                tokens_to_cancel.push(token);
+            }
+        }
+
+        lock_mgr.update_wait_for(update_events);
+        assert_elapsed(
+            move || {
+                for (deadlock_info, lock_info, future) in deadlock_handles {
+                    expect_deadlock(
+                        block_on(future).unwrap(),
+                        deadlock_info.waiter_ts.into(),
+                        lock_info,
+                        Key::from_raw(deadlock_info.key).gen_hash(),
+                        deadlock_info.key,
+                        deadlock_info.wait_chain,
+                    );
+                }
+            },
+            0,
+            250,
+        );
+
+        assert_elapsed(
+            move || {
+                for (lock_info, future) in handles {
+                    expect_key_is_locked(block_on(future).unwrap(), lock_info);
+                }
+            },
+            0,
+            500,
+        );
+    }
+
```

Comment:
Do we have a `upgrading` test case to verify the behaviours are expected when upgrading tikv-servers.

Reply:
It was now manually done.